src/nccl_ofi_net.c

```diff
@@ -1404,84 +1409,75 @@ static ncclResult_t ofi_connect(int dev, void *handle, void **sendComm)
-		NCCL_OFI_TRACE(NCCL_NET, "Connect to self");
-		if (request_to_self != NULL) {
-			NCCL_OFI_WARN("Another connect to self request already in progress");
-			ret = ncclSystemError;
-			goto error;
-		}
-		memcpy(buffer_to_self, local_ep_addr, MAX_EP_ADDR);
-		rc = fi_tsend(sComm->local_ep, buffer_to_self,
-		              MAX_EP_ADDR, NULL, sComm->remote_ep,
-		              sComm->tag | ~max_tag, &req->ctx);
-		if (rc == 0) {
-			NCCL_OFI_TRACE(NCCL_NET, "Connect to self request completed");
-		}
+	do {
+		/*
+		 * TODO: replace it with API of FI_INJECT type when most of
+		 * providers can support it, so that need for completion check
+		 * below can be lifted.
+		 */
+		rc = fi_tsend(sComm->local_ep, conn_info,
+			      sizeof(*conn_info), NULL, sComm->remote_ep,
+			      sComm->tag | ~max_tag, &req->ctx);
+		if (rc == 0)
+			break;
 		else if (rc == -FI_EAGAIN) {
-			NCCL_OFI_TRACE(NCCL_NET, "Connect to self request will be completed once receiver is ready");
-			/* postpone request completion, setting req to NULL prevents freeing it at the end */
-			request_to_self = req;
-			req = NULL;
+			/*
+			 * Process completions so that you have enough
+			 * resources for sending connect message
+			 */
+			ret = nccl_ofi_progress(nccl_ofi_component[dev]);
+			if (OFI_UNLIKELY(ret != 0))
+				goto error;
 		}
 		else {
-			NCCL_OFI_WARN("Unable to send connect message to self for dev %d. RC: %zd, ERROR: %s",
-			              dev, rc, fi_strerror(-rc));
+			NCCL_OFI_WARN("Unable to send connect message for dev %d. RC: %zd, ERROR: %s",
+				     dev, rc, fi_strerror(-rc));
 			ret = ncclSystemError;
 			goto error;
 		}
+	} while (true);
+
+	/* Ensure the message is sent. */
+	/* Skip this step for sending to self */
+	if (connect_to_self) {
+		/* prevent req from freeing */
+		req = NULL;
 	}
 	else {
-		NCCL_OFI_TRACE(NCCL_NET, "Connect to remote");
-		do {
-			/*
-			 * TODO: replace it with API of FI_INJECT type when most of
-			 * providers can support it, so that need for completion check
-			 * below can be lifted.
-			 */
-			rc = fi_tsend(sComm->local_ep, (void *)&local_ep_addr,
-			              MAX_EP_ADDR, NULL, sComm->remote_ep,
-			              sComm->tag | ~max_tag, &req->ctx);
-			if (rc == 0) {
-				break;
-			}
-			else if (rc == -FI_EAGAIN) {
-				/*
-				 * Process completions so that you have enough
-				 * resources for sending connect message
-				 */
-				ret = nccl_ofi_progress(nccl_ofi_component[dev]);
-				if (OFI_UNLIKELY(ret != 0))
-					goto error;
-			}
-			else {
-				NCCL_OFI_WARN("Unable to send connect message for dev %d. RC: %zd, ERROR: %s",
-				              dev, rc, fi_strerror(-rc));
-				ret = ncclSystemError;
-				goto error;
-			}
-		} while (true);
-
-		/* Ensure the message is sent. */
 		do {
 			ret = nccl_ofi_progress(nccl_ofi_component[dev]);
 			if (OFI_UNLIKELY(ret != 0))
 				goto error;
 		} while (req->state != NCCL_OFI_REQ_COMPLETED);
+		free(conn_info);
```

Comment:
We should be freeing up the pointers inside `conn_info` first. Maybe handle this in the exit path?

Reply:
Pointers inside conn_info (namely req) are copies and are freed separately. Please note that conn_info freed conditionally, for remote connections it is freed here in ofi_connect, but for local connections it is freed in ofi_accept.

Reply:
We are freeing `req` later. In general when we are freeing a structure, we free the members first. You can always check for bool `connect_to_self` to identify whether to free an object or not.